### PR TITLE
authenticator-rs: init at 0.7.6

### PR DESCRIPTION
**Pull request by philiptaron**
## Description of changes

[authenticator-rs](https://github.com/grumlimited/authenticator-rs) is a TOTP-MFA application written in Rust and GTK3.

It is currently packaged as a .deb, an .rpm, and is in the [AUR](https://aur.archlinux.org/packages/authenticator-rs-bin).

### Main view

authenticator-rs
authenticator-rs

### Group editing

authenticator-rs

### Account editing

authenticator-rs

### Importing and exporting

Using `yaml` format

    ---
    - name: group name
      url: bbc.co.uk
      entries:
        - label: Account 1
          secret: secret code

authenticator-rs

## Things done

Packaging this application was a little more challenging than I expected.

* The version numbers in the repository didn't match the version numbers in the metadata. The metadata was 0.1 and 0.0.1; the repository is 0.7.6.

* Two tests make use of the network.

* The install process uses a hand-written Makefile that baked in assumptions about `/usr`. It also re-compiled the source.

I also decided to strip "RS" from the `.desktop` name, as it didn't add anything to the functionality of the application.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by philiptaron** [comment]
@matthiasbeyer, as a new committer and a Rust aficionado, perhaps you'd like to review and merge if it passes your bar?

File: pkgs/by-name/au/authenticator-rs/fix-install.patch [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
++
+ 	# Install icons
+ 	$(INSTALL_DATA) data/icons/hicolor/scalable/apps/uk.co.grumlimited.authenticator-rs.svg $(sharedir)/icons/hicolor/scalable/apps/uk.co.grumlimited.authenticator-rs.svg
+ 	$(INSTALL_DATA) data/icons/hicolor/64x64/apps/uk.co.grumlimited.authenticator-rs.64.png $(sharedir)/icons/hicolor/64x64/apps/uk.co.grumlimited.authenticator-rs.png
+@@ -71,7 +69,7 @@ install : release gresource
+ 	# Force icon cache refresh
+ 	touch $(sharedir)/icons/hicolor
+ 
+-	# Install application meta-data
++	# Install application metadata
```
Comment (matthiasbeyer) [inline pkgs/by-name/au/authenticator-rs/fix-install.patch:38]: This patch fragment is not necessary.

File: pkgs/by-name/au/authenticator-rs/fix-install.patch [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+ 	# Install gschema file
+ 	$(INSTALL_DATA) data/uk.co.grumlimited.authenticator-rs.gschema.xml $(sharedir)/glib-2.0/schemas/
+-	
++
+ 	# Install LOCALE files
+ 	rm -fr builddir/
+-	meson builddir --prefix=/usr
++	meson setup --prefix=$(PREFIX) builddir
+ 	DESTDIR=../build-aux/i18n meson install -C builddir
+-	cp -fr build-aux/i18n/usr $(DESTDIR)
```
Comment (matthiasbeyer) [inline pkgs/by-name/au/authenticator-rs/fix-install.patch:53]: Why was this removed?

File: pkgs/by-name/au/authenticator-rs/fix-install.patch [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+diff --git a/README.md b/README.md
+index bd5b29a..9083dce 100644
+--- a/README.md
++++ b/README.md
+@@ -63,7 +63,7 @@ Download from the [release](https://github.com/grumlimited/authenticator-rs/rele
+ 
+ #### Arch Linux
+ 
+-Authenticator RS is also available from [AUR](https://aur.archlinux.org/packages/authenticator-rs-bin/). Or from the release page:
++Authenticator is also available from [AUR](https://aur.archlinux.org/packages/authenticator-rs-bin/). Or from the release page:
```
Comment (matthiasbeyer) [inline pkgs/by-name/au/authenticator-rs/fix-install.patch:84]: This diff is also not necessary.

File: pkgs/by-name/au/authenticator-rs/fix-install.patch [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+diff --git a/data/resources/gtk/ui/main.ui b/data/resources/gtk/ui/main.ui
+index 750f2f6..7ae75cf 100644
+--- a/data/resources/gtk/ui/main.ui
++++ b/data/resources/gtk/ui/main.ui
+@@ -906,7 +906,7 @@
+                 <property name="valign">center</property>
+                 <property name="hexpand">True</property>
+                 <property name="vexpand">True</property>
+-                <property name="label">Authenticator RS</property>
++                <property name="label">Authenticator</property>
```
Comment (matthiasbeyer) [inline pkgs/by-name/au/authenticator-rs/fix-install.patch:97]: IMO this is not necessary. Also it renames the application from what the authors wanted, I would object that.

File: pkgs/by-name/au/authenticator-rs/fix-install.patch [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+               </object>
+               <packing>
+                 <property name="expand">False</property>
+@@ -1111,7 +1111,7 @@
+             <property name="valign">center</property>
+             <property name="hexpand">True</property>
+             <property name="vexpand">True</property>
+-            <property name="label">Authenticator RS</property>
++            <property name="label">Authenticator</property>
```
Comment (matthiasbeyer) [inline pkgs/by-name/au/authenticator-rs/fix-install.patch:106]: Same as above.

File: pkgs/by-name/au/authenticator-rs/fix-install.patch [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+           <packing>
+             <property name="left-attach">0</property>
+diff --git a/data/uk.co.grumlimited.authenticator-rs.desktop b/data/uk.co.grumlimited.authenticator-rs.desktop
+index 1d06a1e..5b939d1 100644
+--- a/data/uk.co.grumlimited.authenticator-rs.desktop
++++ b/data/uk.co.grumlimited.authenticator-rs.desktop
+@@ -1,5 +1,5 @@
+ [Desktop Entry]
+-Name=Authenticator RS
++Name=Authenticator
```
Comment (matthiasbeyer) [inline pkgs/by-name/au/authenticator-rs/fix-install.patch:117]: Same as above.

File: pkgs/by-name/au/authenticator-rs/fix-install.patch [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+diff --git a/src/main.rs b/src/main.rs
+index 3da36b0..4ddacd7 100644
+--- a/src/main.rs
++++ b/src/main.rs
+@@ -82,7 +82,7 @@ fn main() {
+ 
+         gui.set_application(app, connection);
+ 
+-        info!("Authenticator RS initialised");
++        info!("Authenticator initialised");
```
Comment (matthiasbeyer) [inline pkgs/by-name/au/authenticator-rs/fix-install.patch:130]: Same as above.

**Review by matthiasbeyer — CHANGES_REQUESTED** [review]
Hi!

Awesome, thanks for contributing this package!

There are a few things that IMO have to be changed: First of all, the patch file introduces more changes than necessary. Keep the changed stuff from patches as minimal as possible. For example, removing the "RS" prefix - that's renaming the application from what the original authors wanted it to be named. We shouldn't do that.

Also, we might want to see whether we can contribute these changes upstream - what do you think? The changes to the `Makefile` could be contributed upstream I guess... maybe they'll be accepted?

From the nix code itself, this PR looks fine IMO! :tada: Thanks again for taking the time for contributing!

File: pkgs/by-name/au/authenticator-rs/fix-install.patch (hunk 2 above)
Comment (philiptaron) [reply pkgs/by-name/au/authenticator-rs/fix-install.patch]: This is removed in https://github.com/grumlimited/authenticator-rs/pull/138 in order to have the `meson install` line do the copy itself. We need to do that since the current working directory doesn't have `build-aux` when invoked by Nix.

**Comment by philiptaron** [comment]
Thank you so much for the review, @matthiasbeyer. I've opened two PRs against upstream for the issues I ran into while packaging, and the Nixpkgs commit now no longer holds a patch.

> There are a few things that IMO have to be changed: First of all, the patch file introduces more changes than necessary. Keep the changed stuff from patches as minimal as possible. For example, removing the "RS" prefix - that's renaming the application from what the original authors wanted it to be named. We shouldn't do that.

Removed all branding changes. That can stay in my personal NixOS configuration.

> Also, we might want to see whether we can contribute these changes upstream - what do you think? The changes to the `Makefile` could be contributed upstream I guess... maybe they'll be accepted?

Hope so!

> From the nix code itself, this PR looks fine IMO! 🎉 Thanks again for taking the time for contributing!

Thanks for the review, and I think it's ready for a re-review.

**Comment by matthiasbeyer** [comment]
Can you please share a link to the PRs against upstream? Would be nice to track these.

**Comment by philiptaron** [comment]
> Can you please share a link to the PRs against upstream? Would be nice to track these.

Sure thing. They're in the comments, and also here:

1. https://github.com/grumlimited/authenticator-rs/pull/138
2. https://github.com/grumlimited/authenticator-rs/pull/139
